src/audio/multiband_drc/multiband_drc.c

```diff
@@ -0,0 +1,605 @@
+		/* Free all previously allocated blocks in case of an error */
+		if (err < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "crossover_eq_init_coef(), could not assign coeffs to ch %d",
+				    ch);
+			multiband_drc_emphasis_reset_state(state);
+			return err;
+		}
+	}
+
+	comp_cl_info(&comp_multiband_drc, "crossover_eq_init_coef(), initiliazing deemphasis_eq");
+
+	/* Deemphasis: collect the coef array and assign it to every channel */
+	deemphasis = config->deemp_coef;
+	for (ch = 0; ch < nch; ch++) {
+		err = multiband_drc_eq_init_coef_ch(deemphasis, &state->deemphasis[ch]);
+		/* Free all previously allocated blocks in case of an error */
+		if (err < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "crossover_eq_init_coef(), could not assign coeffs to ch %d",
+				    ch);
+			multiband_drc_deemphasis_reset_state(state);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int multiband_drc_drc_init(struct multiband_drc_comp_data *cd, int16_t nch, uint32_t rate)
+{
+	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
+	int num_bands = cd->config->num_bands;
+	struct multiband_drc_state *state = &cd->state;
+
+	int i;
+	int ret;
+
+	/* Allocate all DRC pre-delay buffers and set delay time with band number */
+	for (i = 0; i < SOF_MULTIBAND_DRC_MAX_BANDS; i++) {
+		if (i == num_bands)
+			break;
+
+		comp_cl_info(&comp_multiband_drc, "drc_init(), initiliazing drc band %d", i);
+
+		ret = drc_init_pre_delay_buffers(&state->drc[i], (size_t)sample_bytes, (int)nch);
+		if (ret < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "drc_init(), could not init pre delay buffers");
+			return ret;
+		}
+
+		ret = drc_set_pre_delay_time(&state->drc[i],
+					     cd->config->drc_coef[i].pre_delay_time, rate);
+		if (ret < 0) {
+			comp_cl_err(&comp_multiband_drc,
+				    "drc_init(), could not set pre delay time");
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int multiband_drc_setup(struct multiband_drc_comp_data *cd, int16_t channels, uint32_t rate)
+{
+	int ret;
+
+	/* Reset any previous state */
+	multiband_drc_reset_state(&cd->state);
+
+	/* Setup Crossover, Emphasis EQ, and Deemphasis EQ */
+	ret = multiband_drc_crossover_eq_init_coef(cd, channels);
+	if (ret < 0)
+		return ret;
+
+	/* Setup Multiband DRC */
+	ret = multiband_drc_drc_init(cd, channels, rate);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+/*
+ * End of Multiband DRC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *multiband_drc_new(const struct comp_driver *drv,
+					  struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct multiband_drc_comp_data *cd;
+	struct sof_ipc_comp_process *ipc_multiband_drc =
+		(struct sof_ipc_comp_process *)comp;
+	size_t bs = ipc_multiband_drc->size;
+	int ret;
+
+	comp_cl_info(&comp_multiband_drc, "multiband_drc_new()");
+
+	/* Check first before proceeding with dev and cd that coefficients
+	 * blob size is sane.
+	 */
+	if (bs > SOF_DRC_MAX_SIZE) {
```

Comment:
do we have a min size?

Reply:
In fact I don't quite know about this checker. My understanding is not to check the blob data size is correct, just to check the blob data is not insane. Previous components always check Max size only. Do we need to check min?

Reply:
I've had this check in components exactly for the reason to prevent insane big blobs to cause the firmware to fail. The too small blob could fail similarly or via parameters sanity check, as you prefer.

Reply:
Yea, I expect the fuzzer to break this pretty quickly if we are only doing max checks. I am right now working to give the fuzzer access to all of the modules.